### PR TITLE
Allow injecting API keys

### PR DESCRIPTION
**Pull request by pbugnion**
This PR allows injection of a Google Maps API key. This increases the number of requests that a user can make, and may even be necessary for some services. This addresses issue #61 .

The interface is:

```python

import gmaps

gmaps.configure(api_key="your-api-key")

m = gmaps.Map()
d = gmaps.Directions(data=[[45.0, 5.0], [46.0, 7.0], [47.0, 6.0]])

m.add_layer(d)
m
```

To unset the API key, use:

```python
gmaps.configure(api_key=None)
```

The configuration can be applied at any time, but any existing widgets will have to be reloaded (by re-executing the cell in which they are rendered) for the configuration changes to affect that widget.

### CODE REVIEW

**Comment by pbugnion** [comment]
This is based off PR #60, so should be merged after that. All that is left to do is document the `maps.configure` function, I think.

**Comment by pbugnion** [comment]
@xoolive If you have time to test this out, that would be very welcome.

**Comment by xoolive** [comment]
Let me try that early next week 👍

File: js/src/jupyter-gmaps.js [hunk 1]
```diff
@@ -42,16 +61,16 @@ var DirectionsLayerView = GMapsLayerView.extend({
             origin: orig,
             destination: dest,
             waypoints: wps,
-            travelMode: google.maps.DirectionsTravelMode.DRIVING
+            travelMode: google.maps.TravelMode.DRIVING
```
Comment (xoolive) [inline js/src/jupyter-gmaps.js]: Isn't it `google.maps.DirectionsTravelMode.DRIVING`? (just asking)

Comment (pbugnion) [inline js/src/jupyter-gmaps.js]: Well spotted. It actually looks like both are valid, though the documentation only mentions `TravelMode`, I think:

https://developers.google.com/maps/documentation/javascript/3.exp/reference#TravelMode

**Comment by xoolive** [comment]
Unfortunately, this doesn't work. :cry: 

`"Google Maps API error: MissingKeyMapError https://developers.google.com/maps/documentation/javascript/error-messages#missing-key-map-error" js:35:179`

Well, it seems that it sometimes still works without an API key, but here is the result of my experiments:
- import gmaps; gmaps.Map() => sometimes works
- import googlemaps; googlemaps.Client("AI....."); import gmaps; gmaps.Map() => `MissingKeyMapError`
- import gmaps; gmaps.configure(api_key="AI.....") => `MissingKeyMapError`
- import gmaps; gmaps.configure(api_key=None) => `MissingKeyMapError`

Note that during my tests for my quickfix, if I entered an invalid key or a key not allowed for Javascript API, I had a different error.

File: js/src/jupyter-gmaps.js (hunk 1 above)
Comment (xoolive) [inline js/src/jupyter-gmaps.js]: Looks reasonable! :smile:

**Comment by pbugnion** [comment]
Thanks a lot for checking. It's strange that you're getting an error. It's definitely working for me. I even tried in a fresh virtual machine:

screen shot 2016-06-27 at 19 25 46

Do you mind possibly checking again from a new notebook? To make checking easier, I've created release 0.3.1.rc1 which you can install with:

```
$ pip uninstall gmaps
$ pip install gmaps==0.3.1rc1
$ jupyter nbextension enable --py gmaps
```

**Comment by xoolive** [comment]
Thank you for the trouble!
So, on one of my computer (for the other one, you will have to wait few days...):
- your example works like a charm in a clean notebook;
- in my big notebook (for students... 😜 ), the following cell fails at first execution (`MissingKeyMapError`), but works when I execute the code again. (don't ask...)

```python
import gmaps
#print(gmaps.__version__)
m = gmaps.Map()
gmaps.configure(api_key="AIza...")
#gmaps.configure(api_key=None)

data = [(48.85341, 2.3488), (50.85045, 4.34878), (52.37403, 4.88969)]
directions_layer = gmaps.Directions(data=data)
m.add_layer(directions_layer)
m
```

Also note that the `pip uninstall` does not clean everything. I spotted the `share/jupyter/nbextensions/jupyter-gmaps` still present and containing the three js files...

**Comment by pbugnion** [comment]
Thanks a lot for checking this. I'll work on releasing `0.3.1` with the ability to inject API keys. 

I think that the problem with existing notebooks comes from `GoogleMapsLoader` (the package that we use to load Google maps in requireJS) being a bit of a hack. It manually injects `<script>` tags in the html. These tags persist until you close the server *and* the browser tab in which the notebook is open. 

And yes, the installation and uninstallation of notebook widgets through pip is still fairly broken. It also doesn't really respect virtual environments (for the JavaScript part, for instance). I'm hoping it will get fixed in the next releases of [ipywidgets](https://github.com/ipython/ipywidgets).

**Pull request merged**